ot/gromov.py

```diff
@@ -658,21 +723,38 @@ def fused_gromov_wasserstein(M, C1, C2, p, q, loss_fun='square_loss', alpha=0.5,
     def f(G):
         return gwloss(constC, hC1, hC2, G)
 
-    def df(G):
-        return gwggrad(constC, hC1, hC2, G)
+    if symmetric:
+        def df(G):
+            return gwggrad(constC, hC1, hC2, G)
+    else:
+        constCt, hC1t, hC2t = init_matrix(C1.T, C2.T, p, q, loss_fun)
+
+        def df(G):
+            return 0.5 * (gwggrad(constC, hC1, hC2, G) + gwggrad(constCt, hC1t, hC2t, G))
 
+    if loss_fun == 'kl_loss':
+        armijo = True
+
+    # cg for GW is implemented using numpy on CPU
+    np_ = NumpyBackend()
+    if armijo:
+        def line_search(cost, G, deltaG, Mi, f_G, **kwargs):
+            return line_search_armijo(cost, G, deltaG, Mi, f_G, nx=np_, **kwargs)
+    else:
+        def line_search(cost, G, deltaG, Mi, f_G, **kwargs):
+            return solve_gromov_linesearch(G, deltaG, f_G, C1, C2.T, M=(1 - alpha) * M, reg=alpha, nx=np_, **kwargs)
     if log:
-        res, log = cg(p, q, (1 - alpha) * M, alpha, f, df, G0, armijo=armijo, C1=C1, C2=C2, constC=constC, log=True, **kwargs)
+        res, log = cg(p, q, (1 - alpha) * M, alpha, f, df, G0, line_search, log=True, **kwargs)
         fgw_dist = nx.from_numpy(log['loss'][-1], type_as=C10)
         log['fgw_dist'] = fgw_dist
         log['u'] = nx.from_numpy(log['u'], type_as=C10)
         log['v'] = nx.from_numpy(log['v'], type_as=C10)
         return nx.from_numpy(res, type_as=C10), log
     else:
-        return nx.from_numpy(cg(p, q, (1 - alpha) * M, alpha, f, df, G0, armijo=armijo, C1=C1, C2=C2, constC=constC, **kwargs), type_as=C10)
+        return nx.from_numpy(cg(p, q, (1 - alpha) * M, alpha, f, df, G0, line_search, log=False, **kwargs), type_as=C10)
 
 
-def fused_gromov_wasserstein2(M, C1, C2, p, q, loss_fun='square_loss', alpha=0.5, armijo=False, G0=None, log=False, **kwargs):
+def fused_gromov_wasserstein2(M, C1, C2, p, q, loss_fun='square_loss', symmetric=None, alpha=0.5, armijo=False, G0=None, log=False, **kwargs):

```

Comment:
same her the function should call fused_gromov_wasserstein and just set the gradients